### DISCUSSION TITLE
New Backend

### DISCUSSION THREAD

**Discussion by kstan79** [bpmnServer/bpmn-server#123] [General]
I know that current backend for prototyping and not design for production use.

Since bpmn-server is pure backend engine I recommend to create new backend using nestjs. it offer benefit which included:
1. More robust design
2. bigger ecosystem and we able to use lot of nestjs plug-in like sso authentication, swagger-ui, middleware
3. We have better way to manage complex workflow requirement
4. Use openapi we no need maintain client library. It mean we can accept frontend request from Java,python, .net without headache 

Others bpmn project like kogito, they did similar and are they use quarkus which can allow developer use plenty of quarkus plugin. 

I use it sometimes but I can't stay with them cause it is too rigit 

If you agree we may work on it

**Comment by bpmn-ts** [comment]
Hi
Sounds great
Please go ahead by providing more description and design information
Thanks

Sent from my iPhone

On Sep 21, 2023, at 6:44 AM, Ks Tan ***@***.***> wrote:

﻿

I know that current backend for prototyping and not design for production use.

Since bpmn-server is pure backend engine I recommend to create new backend using nestjs. it offer benefit which included:

  1.  More robust design
2 bigger ecosystem and able to use lot of nestjs plug-in like sso authentication, swagger-ui, middleware
  2.  We have better way to manage complex workflow requirement

So for bpmn project like kogito they are using quarkus which can allow developer use plenty of quarkus plugin

If you agree we may work on it

**Comment by kstan79** [comment]
We may start from easier way like:



1. Create an kickstart git repo like "bpmn-nest-template"
2. Inside content latest nestjs project and appropriate route/service controller
3. Put some bpmn sample, delegates inside
4. Write read me for how to kick start:
- add more process definition
- change configuration via .env
- try api via swagger
- turn on api key or sso jwt token (like keycloak)
- how to build for production
- and etc

One comfortable we may create cli executor and user install it via "npm install -g bpmn-server-cli"


Then setup and maintain the project using
- bpmn-server-cli init my-proj
- bpm-server-cli start/stop
- bpmn-server-cli add-deligate uploads3 emailuser push-notification 
- etc

**Comment by ralphhanna** [comment]
Great
Please go ahead with a separate branch

I will be done with a full release in couple of days
Thanks

Sent from my iPhone

On Sep 21, 2023, at 6:15 PM, Ks Tan ***@***.***> wrote:

﻿

We may start from easier way like:

  1.  Create an kickstart git repo like "bpmn-nest-template"
  2.  Inside content latest nestjs project and appropriate route/service controller
  3.  Put some bpmn sample, delegates inside
  4.  Write read me for how to kick start:

  *   add more process definition
  *   change configuration via .env
  *   try api via swagger
  *   turn on api key or sso jwt token (like keycloak)
  *   how to build for production
  *   and etc

One comfortable we may create cli executor and user install it via "npm install -g bpmn-server-cli"

Then setup and maintain the project using

  *   bpmn-server-cli init my-proj
  *   bpm-server-cli start/stop
  *   bpmn-server-cli add-deligate uploads3 emailuser push-notification
  *   etc

**Comment by ingpconci** [comment]
I tried to create a loopback extension to use with the loopback backend framework
https://github.com/ingpconci/loopback4-bpmn-server
I know the loopback project is not updated like Nestjs, but the project is hosted on the OpenJS foundation
https://openjsf.org/blog/2022/04/25/node-js-framework-loopback-graduates-to-at-large-project-at-openjs-foundation/
What do you think about it?

**Comment by kstan79** [comment]
Sounds good, I will spend sometime check loop back is it good for me,

**Comment by ralphhanna** [comment]
Can we think of Bpmn-server as a microservice that can be used in either framework?

**Comment by kstan79** [comment]
IMHO bpmn act as purely backend microservices, it doesn't effect much as long as it comply openapi api standard. All languages and framework can tan to openapi easily.

I not recommend bpmn server embed as library to support 2 or more framework cause it add burden to development for testing and documentation. Let's pick one and stick with it until perfect, than decide again.

In typescript I only have experience in nuxt and nestjs, never try loopback framework so I can't give comment about loopback good or bad yet.

Nest has build plenty of features I think suitable for industry grade application, bpmn-server adopt by industry most of the time so I feel we park inside good for us

**Comment by ralphhanna** [comment]
Hi
I have looked at https://github.com/ingpconci/loopback4-bpmn-server
and it seems to be a full port to loopback4 which will force everyone to use loopback4.

I am hoping that only the bpmn-client will be ported to lb4 with is minimal code while keeping bpmn-server as pure nodeJS

Thanks, Let us keep the discussion going

**Comment by ingpconci** [comment]
Hi
I use the loopback as a client for bpmn-server, because I want to use the loopback API framework to control authentication, filtering and connecting to the relational database PostgreSQL.
I imported the library, created the server instance and used it in the controllers.
I don't modify the code of bpmn-server and I change only the 4 customization files for the library

![image](https://github.com/ralphhanna/bpmn-server/assets/8488830/d3391dbc-077c-40e6-929c-bc952b5efcae)

The main effort is focused on defining a flexible system user task assignment; to facilitate the acceptance of BPMN workflow, I use this approach based on association between Lane in diagram and Use Role

![image](https://github.com/ralphhanna/bpmn-server/assets/8488830/75e83f02-dc30-4399-b692-c44f1d430cf2)

In the extension, I put the controller API endpoints to support the Role management.
In the picture there is the Html interface.
![image](https://github.com/ralphhanna/bpmn-server/assets/8488830/3d894bfe-bd67-4569-8e53-876209e76ed4)

I would like to put in the loopback bpmn-server extension an API endpoint that returns the information about the User Task Form.
At the moment, I define the BPMN process with camunda modeler and use the camunda form definition.

![image](https://github.com/ralphhanna/bpmn-server/assets/8488830/7bd58c18-5a20-497c-9c04-949dc0d08d7f)

`
<bpmn:process id="Process_07ii97u" isExecutable="true">    <bpmn:extensionElements>      
<zeebe:userTaskForm id="userTaskForm_08ui781">{  "components": [    {      "text": "&lt;h2&gt;(1.1) New nonconformity&lt;/h2&gt;",      "label": "Text view",
…..
}</zeebe:userTaskForm>
`

In my Html App, I have developed a form engine that uses this information in the model definition to build the Form, but in the future I would like to facilitate the construction of the form at server side. The function must provide a form representation that permit to build a non Html form (e.g. a standalone C++ app form engine)

![image](https://github.com/ralphhanna/bpmn-server/assets/8488830/234f3c78-4340-439d-bb09-ad7c11844421)

**Comment by bpmn-ts** [comment]
Thanks very much for the explanation, that looks great, the only challenge
I found is how to update bpmn-server?

Is there a way to use npm instead of copying the code into the library
folder?

Please let me know if you need any changes from my end.

Thanks

**Comment by kstan79** [comment]
After view back your project seems u did extension for loopback only. Previously misunderstand and thought you host whole loopback application in repo.

@bpmn-ts you may need to decide is loopback a good framework and qualify as firstclass citizen for bpmn or not. if yes we may update documentation according that approach. Else, i will spend times to do in nestjs.

@ingpconci  can you make your effort become npm package so it more portable? I'm excited actually because your may solve a big issue which can make bpmn-server use in production

**Comment by ingpconci** [comment]
I clone the project locally

` git clone https://github.com/ralphhanna/bpmn-server.git`

Then compile and pack
```
> npm run clean
> npm run build
> npm pack
...
npm notice === Tarball Details ===
npm notice name:          bpmn-server
npm notice version:       1.3.28
npm notice filename:      bpmn-server-1.3.28.tgz
npm notice package size:  10.2 MB
npm notice unpacked size: 49.3 MB
npm notice shasum:        7c52e0ac88d9b340516c8b484db5d7700af73c26
npm notice integrity:     sha512-DqC9SRCt3HlzT[...]mrIoswkvWnNjw==
npm notice total files:   851
npm notice
bpmn-server-1.3.28.tgz
```
After that I insert in the package.json

```
 "dependencies": {
    "bpmn-moddle": "^8.0.1",
    "bpmn-server": "file:bpmn-server-1.3.28.tgz",
    "camunda-bpmn-moddle": "^7.0.1",
    "tslib": "^2.0.0"
  },
```

**Comment by ingpconci** [comment]
To use in the main loopback application, I import these libraries into the application.ts

```
    "loopback-connector-postgresql": "^6.0.8",
    "loopback4-tenant-table-filter": "file:loopback4-tenant-table-filter-1.0.8.tgz",
    "bpmn-server": "file:bpmn-server-1.3.28.tgz",
    "loopback4-bpmn-server": "file:loopback4-bpmn-server-1.0.0.tgz"
```

and insert the initialization code

```
    //--------------------------------------------------------------------------
    // BpmnServerComponent
    //--------------------------------------------------------------------------
    const definitionsPath = path.join(__dirname, '../processes/')
    console.log('AppServerApplication:startup: bpmnProcessDefinitionDirectory=', definitionsPath);
    const componentOptions: Loopback4BpmnServerComponentOptions = {
      bpmnProcessDefinitionDirectory: definitionsPath
    }
    this.configure(Loopback4BpmnServerComponentBindings.COMPONENT).to(componentOptions);
    this.component(Loopback4BpmnServerComponent);
```

Remember that the extension uses a multitenant authorization extension
https://github.com/ingpconci/loopback4-tenant-table-filter
that need to be configured before the BpmnServerComponent

```
    //--------------------------------------------------------------------------
    // Multitenant Authorization from loopback4-tenant-table-filter
    // Authorization provider
    //--------------------------------------------------------------------------
    const authorizationOptions: AuthorizationOptions = {
      precedence: AuthorizationDecision.DENY,
      defaultDecision: AuthorizationDecision.DENY,
    };
    this.configure(AuthorizationBindings.COMPONENT).to(authorizationOptions);
    this.component(AuthorizationComponent);
    // bind the authorizer provider
    this.bind('authorizationProviders.multitenant-authorizer-provider')
      .toProvider(MultitenantAuthorizerProvider)
      .tag(AuthorizationTags.AUTHORIZER);
```
The loopback4-tenant-table-filter permit to have a unique DB but different isolated organization
For example, the bpmnProcessInstance table contains records of different tenant
![image](https://github.com/ralphhanna/bpmn-server/assets/8488830/40b57aa5-a58e-4051-9725-79b21e236eb7)

When I start the process I set this instance variable in the process with the tenantId of the user

```
      inputData: {
        tenantId: this.authorizationService.currentTenantId,
        startProcessUserId: this.currentUser.id
      }
    const startEngineResult = await this.bpmnServer.engine.start(processName, inputData);
```

The bpmn server engine use this  information for all operation in the DB in the custom datastore that uses the 
bpmnProcessInstanceRepository.createNoTenantIdFilter


```
      
  // save instance to DB
  static seq = 0;
  private async saveInstance(instance, items) {
    //	this.logger.log("Saving...");
    console.log('CustomDataStore.saveInstance:');
......
......
      //-----------------------------------------------------------
      //Get tenantId from inputData
      //-----------------------------------------------------------
      const currentTenantId = instance.data.tenantId;
      bpmnProcessInstance.tenantId = currentTenantId;

      const result = await this.server.configuration.database.loopbackRepositories.bpmnProcessInstanceRepository.createNoTenantIdFilter(bpmnProcessInstance);
      debug('saveInstance: new instance inserted! result=', result);
```

**Comment by ralphhanna** [comment]
I believe the full App (Web UI, Server) can use any framework. but I would rather have bpmn-server not rely on any framework, because that would be restrictive.

In other words  I like  that bpmn-server to remain on pure nodeJS and only offer ExpressJS for demo purposes only?

So the question is what features in bpmn-server can be enhanced using these frameworks?

Any thoughts?

**Comment by kstan79** [comment]
I don't think it is good ideal to maintain as independent express due to below reason

1. Pure express framework far behind production use, there is a lot of thing need to do and until up to production level to use. Such as cluster, cache, security, queue system and etc which need high experience developer assemble it in express. Use 3rd party framework provide reliable documentation and step to do that.

2. I believe you wish to remain independent with good reason, but I think that make project getting far from provide easy to use bpmn solution. my thought is bpmn his kind of low code solution and everything include setup/database shall implement low code concept as well. I feel you pick mongodb is wise decision cause it also match low code philosophy concept.

Current project quality, completeness, reliability yet to achieve perfect. In this situation flexible is kind of no-no and drag the progress. Why don't Make everything static at the time being until near perfect, then decide again in future?

**Comment by kstan79** [comment]
Anyhow, it is my personal perspective, which I believe people wish to find robust bpmn framework rather than bpmn library.

I'm ok with using someone bpmn framework, as long as reliable and suitable

**Comment by ralphhanna** [comment]
Not sure I understand your recommendation, can you please clarify what do you recommend, keep in mind Express is only used for demo purposes

**Comment by kstan79** [comment]
1. Many kind users, some is programmer and some is not. 
2. Event though the specific user is programmer, he might not expert in typescript and express
3. The user may feel interested in this project and wish to implement in their organisation
4. They may not have capability to walk through complex setup process, and they wish to have something just work fine for production
5. Most of the time, their concern is how to put bpmn inside, how to chain this microservice to their existing webapp. They not care what framework it is, they care is reliable and well documented step. If possible it can install and forget.


Technically most of time the users want something prepackaged for production use, included me. For production package, it need:
1. quick start
2. Documentation for implementation, configuration base on common needs:
1. Authentication/authorisation
2. Database setup
3. Integration with existing frontend web app
3. Some test case, some video guides or walk through


Some of above requirement will need the project to settle down best practice of implement bpmn, in this case flexibility may not a good ideal. We may rather put that as advance topic like how to use others framework (if it is necessary)

**Comment by ingpconci** [comment]
When I chose the bpmn-server library, I checked other libraries and full frameworks like Camunda Products.
The Camunda product meets all of your needs, but it is not open source.
I agree with @ralphhanna. It's better to keep the core as a pure Node.js library.
Now there is an Express client for demo purposes to verify the behavior of the library, but I agree that it should not be used for production.
My approach was to use the library in the main application service based on loopback.
Here are the advantages of my approach:
- The BPMN data is saved in the same database as the main application server.
- We don't have to manage any other services or Java applications on the server.
However, there is one downside:
- The main application cannot be scaled because it is currently not possible to have two instances of the engine that share process instance data.

I agree that a reference framework could benefit the development of the bpmn-server library. I am also open to discussing the possibility of making my main loopback application service, along with the HTML angular client, open source. This could be useful for managing and interacting with the bpmn server.

**Reply by ralphhanna** [reply]
Thanks for the feedback, I have enhanced bpmn-server to be able to run in multiple instance of the engine and share the same DB.

Hope this solves your concern about the downside.

Also, in the upcoming release I will support multi-tenants

**Reply by kstan79** [reply]
This nice

**Comment by ralphhanna** [comment]
@ingpconci Thanks very much for you comments and contribution.

I definitely appreciate any additional contribution to enhance the Client UI.

As for the server, would a package like Memcached help solve the scalability issue?

Thanks again

**Comment by ralphhanna** [comment]
@ingpconci , I also noticed from your code of support mySQL is using an older version of the code, and I don't see you saving tokens into the instance.

I can support mySQL as follows:
- Instance data is saved as json
- Can search on instances (findInstance) or items (findItems) using instance columns or item columns
- Will also store user assignment (role) for the item
- Option to store tenantId

**Comment by ingpconci** [comment]
@ralphhanna , I updated the bpmn-server on my installation with the last version, and I see that there is a new Class InstanceLocker.
How I can modify this Locker class in order to use my custom files
![image](https://github.com/bpmnServer/bpmn-server/assets/8488830/241c0b90-ae75-4539-98d3-7112971b6e2b)

I need to change the DefaultConfiguration.ts and add a method to permit a custom locker?

During the first test , I've found a bug in the Definition.ts:107:34.
When a lane is empty, like this
![image](https://github.com/bpmnServer/bpmn-server/assets/8488830/7d67fd1d-9e9e-4eb0-902f-36c6ad01f485)
the loading sent an error
***************BPMNServer UNCAUGHT ERROR*********** TypeError: Cannot read properties of undefined (reading 'forEach')

`ls.lanes.forEach(lane=>{
                    lane.flowNodeRef.forEach(fnr=>{
                            let target = this.getNodeById(fnr.id);
                            target.lane=lane.name;
                        });
                });
            });`

**Reply by ingpconci** [reply]
Regarding the customLocker I created and used a new class in my custom files.
![image](https://github.com/bpmnServer/bpmn-server/assets/8488830/0b12563f-05e1-4cd9-86e9-dc3c10086f78)

**Comment by ralphhanna** [comment]
The intent is instance_locker be private to the dataStore

**Reply by ingpconci** [reply]
Yes, it is correct.

**Comment by kstan79** [comment]
Nestjs using `@inject`, which cause us hard to integrate bpmn-server perfectly. 
After lot of tries, the reliable way to implement `bpmn-server` is use 'Event/Listener routing' concept:

1. initialize bpmn-server in new workflow.service class
2. listen all bpmn-listener event using `applyEventListener` from `workflow.service`, and route the event base on own preference (refer source code)
3. any event happen in bpmn-server, will auto emit new nestjs's event, then we can listen nestjs event everywhere.
4. during user call controller to `invoke` or `begin` workflow, will trigger nestjs event, we route it back to bpmn-server.

Below is the `workflow.service` I did.

```ts
/**
 * This file execute workflow bpmn-server command
 * This file was automatically generated by simpleapp generator.
 * DONT CHANGE THIS FILE !!
 * MODIFICATION OVERRIDE BY GENERATEOR
 * last change 2023-10-28
 * Author: Ks Tan
 */
import { WorkflowConfig } from './workflow.config';
import {
  BadRequestException,
  Injectable,
  InternalServerErrorException,
  Logger,
  NotFoundException,
} from '@nestjs/common';
import {
  WorkflowSettingApiSchema,
  WorkflowDataApiSchema,
  WorkflowProcess,
  WorkflowTask,
  UserTaskActors,
  UserTaskType,
  ServiceTaskData,
  UserTaskData,
  UserTaskEventType,
  ServiceTaskEventType,
} from './workflow.type';
import { EventEmitter2, OnEvent } from '@nestjs/event-emitter';
import {
  BPMNServer,
  Logger as bpmnlogger,
  BPMN_TYPE,
  EXECUTION_EVENT,
} from 'bpmn-server';
// import { configuration } from './configuration';
import { UserContext } from '../commons/user.context';
//import * as formschemas from '../../workflow/formschema';
export {
  WorkflowSettingApiSchema,
  WorkflowDataApiSchema,
} from './workflow.type';

@Injectable()
export class WorkflowService {
  protected bpmnServer: BPMNServer;
  private readonly logger = new Logger();
  constructor(
    private workflowconfig: WorkflowConfig,
    private eventEmitter: EventEmitter2,
  ) {
    this.bpmnServer = new BPMNServer(this.workflowconfig.getConfig());
    this.bpmnServer.listener.on(
      'all',
      async (event) => await this.applyEventListener(event),
    );
  }

  /**
   * get all workflow userTask for specific uid (uuid)
   * @param appuser
   * @returns array of usertask
   */
  async getActorUserTask(appuser: UserContext, uid: string) {
    const tmp: any[] = await this.bpmnServer.dataStore.findItems({
      'items.status': 'wait',
      'data.tenantId': appuser.getTenantId(),
      //filter uid properties base on function
      $or: [
        { 'items.assignee': appuser.getUname() },
        { 'items.candidateUsers': appuser.getUname() },
        { 'items.candidateGroups': appuser.getGroup() },
      ],
    });
    const result: UserTaskType[] = tmp.map((proc) => {
      const subitem: UserTaskType = {
        id: proc.instanceId,
        taskId: proc.id,
        elementId: proc.elementId,
        name: proc.name,
        processName: proc.processName,
        assignee: proc.assignee,
        candidateUsers: proc.candidateUsers,
        candidateGroups: proc.candidateGroups,
        data: proc.data,
        vars: proc.vars,
        startedAt: proc.startedAt,
        followUpDate: proc.followUpDate,
        dueDate: proc.dueDate,
        priority: proc.priority,
        type: proc.type,
        userId: proc.userId,
      };
      return subitem;
    });
    return result;
    // throw new BadRequestException(`getActorUserTask is not supported yet`)
  }
  /**
   * get all workflow userTask for my uid (uuid)
   * @param appuser
   * @returns array of usertask
   */
  async getMyUserTask(appuser: UserContext) {
    return await this.getActorUserTask(appuser, appuser.getUid());
  }

  /**
   * !! waiting founder
   * deveploper method
   * create new process definitions
   */
  async newWorkflowDefinitions(
    appuser: UserContext,
    workflowName: string,
    xml: string,
  ) {
    throw new BadRequestException(
      `newWorkflowDefinitions "${workflowName}" is not supported yet`,
    );
  }

  async getWorkflowDefinitions(appuser: UserContext) {
    throw new BadRequestException(
      `getWorkflowDefinitions is not supported yet`,
    );
  }

  async readWorkflowDefinition(appuser: UserContext, workflowname: string) {
    throw new BadRequestException(
      `readWorkflowDefinition is not supported yet`,
    );
  }

  async deleteWorkflowDefinition(appuser: UserContext, workflowname: string) {
    throw new BadRequestException(
      `deleteWorkflowDefinition is not supported yet`,
    );
  }

  async updateWorkflowDefinition(
    appuser: UserContext,
    workflowname: string,
    xml: string,
  ) {
    throw new BadRequestException(
      `updateWorkflowDefinition is not supported yet`,
    );
  }

  /**
   * admin workflow service, obtain all active workflows and it's pending task from all tenant
   * status: done
   * @param appuser
   * @returns array of workflow process and it's waiting task
   */
  async getAllActiveProcesses(appuser: UserContext) {
    const alldata = await this.bpmnServer.dataStore.findInstances(
      { status: 'running' },
      'summary',
    );
    const result = this.workflowProcessArrayMaping(alldata);
    return result;
  }

  /**
   * delete a workflow process from database
   * @param appuser
   * @param id
   * @returns
   */
  async deleteWorkflowById(appuser: UserContext, id: string) {
    if (await this.getWorkflowById(appuser, id)) {
      const result = await this.bpmnServer.dataStore.deleteInstances({
        id: id,
      });
      return {
        status: 'ok',
        deletedCount: result['deletedCount'],
      };
    }
  }

  /**
   * invoke workflow process to next step
   * @param appuser
   * @param id
   * @param taskId
   * @param data
   * @returns
   */
  async invokeWorkflowUserTask(
    appuser: UserContext,
    taskId: string,
    data: any,
  ) {
    try {
      const response = await this.bpmnServer.engine.invoke(
        { 'items.id': taskId, 'items.status': 'wait' },
        data,
        appuser.getUid(),
        { appuser: appuser },
      );
      return {
        parentKeys: Object.keys(response),
        parentId: response.id,
        data: response.instance.data,
        // parentId:response,
        input: response.item.input,
        vars: response.item.vars,
        id: response.item.id,
        status: response.item.status,
      };
    } catch (e) {
      throw new NotFoundException(e);
    }
  }

  /**
   * !! waiting founder
   * use in adhoc situation override bpmn assignee/candidateUsers/Groups with data
   * status: draft
   * @param appuser
   * @param id
   * @param taskId
   * @param data
   */
  async updateWorkflowUserTask(
    appuser: UserContext,
    id: string,
    taskId: string,
    assignmentdata: UserTaskActors,
  ) {
    try {
      const result = await this.bpmnServer.dataStore.findInstance(
        { id: id },
        {},
      );

      const usertask = result.items.find((item) => item.id == taskId);
      if (usertask) {
        const result = await this.bpmnServer.engine.assign(
          { id: id, 'items.id': taskId },
          { newData: 1 }, //some data want to put in
          assignmentdata,
          appuser.getUid(),
          {}, //options
        );
        return 'OK'; //Object.keys(result)
      } else {
        throw new NotFoundException(`user task id:${taskId} not found`);
      }
    } catch (e) {
      throw new NotFoundException(e);
    }

    // throw new BadRequestException(`updateWorkflowUserTask is not supported yet`)
  }

  /**
   *
   * @param appuser search all workflow in current tenant
   * @param filter
   * @returns
   */
  async searchRunningWorkflowProcesses(appuser: UserContext, filter: any) {
    // filter['data.tenantId']= appuser.getTenantId()  //tmp disabled waiting
    filter['status'] = 'running';
    return await this.searchWorkflowProcesses(appuser, filter);
  }

  /**
   *
   * @param appuser search all workflow in current tenant
   * @param filter
   * @returns
   */
  async searchWorkflowProcesses(appuser: UserContext, filter: any) {
    filter['data.tenantId'] = appuser.getTenantId(); //tmp disabled waiting
    const alldata = await this.bpmnServer.dataStore.findInstances(
      filter,
      'summary',
    );
    const result = this.workflowProcessArrayMaping(alldata);
    return result;
  }

  async getPendingProcess(appuser: UserContext) {
    const tmp: any[] = await this.bpmnServer.dataStore.findItems({
      'items.status': 'wait',
    });

    const result = tmp.map((proc) => ({
      id: proc.instanceId,
      assignee: proc.assignee,
      candidateUsers: proc.candidateUsers,
      candidateGroups: proc.candidateGroups,
      data: proc.data,
      processName: proc.processName,
      startedAt: proc.startedAt,
      type: proc.type,
      taskName: proc.name,
      userId: proc.userId,
    }));
    return result;
  }

  /**
   * initialise new workflow process, also listening event 'workflow.start'
   * status: done
   * @param appuser
   * @param workflowName
   * @param data
   * @returns workflowsummary
   */
  @OnEvent('workflow.start')
  async startWorkflow(appuser: UserContext, workflowName: string, data?: any) {
    try {
      // console.log('startWorkflow started: ', workflowName);
      data.tenantId = appuser.getTenantId();
      data.orgId = appuser.getOrgId();
      data.branchId = appuser.getBranchId();
      const result = await this.bpmnServer.engine.start(
        workflowName,
        data,
        undefined,
        appuser.getUid(),
        {
          appuser: appuser,
        },
      );

      // console.log("startWorkflowstartWorkflowstartWorkflow",result)
      return {
        id: result.id,
        name: result.name,
        status: result.status,
        userId: result.userName,
        startedAt: result.item.startedAt,
      };
    } catch (e) {
      throw new BadRequestException(e, JSON.stringify(data));
    }
  }

  /**
   * read single process full property
   * @param appuser
   * @param id
   * @returns
   */
  async getWorkflowById(appuser: UserContext, id: string) {
    try {
      const data = await this.bpmnServer.dataStore.findInstance({ id: id }, {}); // .findInstances({id:id},'full')
      const result = this.workflowProcessMaping(data);
      return result;
    } catch (e) {
      throw new NotFoundException(e);
    }
  }
  async getPendingWorkflows(appuser: UserContext) {
    const tmp: any[] = await this.bpmnServer.dataStore.findItems({
      'items.status': 'wait',
    });

    const result = tmp.map((proc) => ({
      id: proc.instanceId,
      assignee: proc.assignee,
      candidateUsers: proc.candidateUsers,
      candidateGroups: proc.candidateGroups,
      data: proc.data,
      processName: proc.processName,
      startedAt: proc.startedAt,
      type: proc.type,
      taskName: proc.name,
      userId: proc.userId,
    }));
    return result;
  }

  /**
   * obtain specific bpmn task camunda:formKey
   * @param appuser
   * @param processName
   * @param elementId
   * @returns properties
   */
  async getWorkflowUserTaskFormKey(
    appuser,
    processName: string,
    elementId: string,
  ) {
    try {
      let definition = await this.bpmnServer.definitions.load(processName);
      let node = definition.getNodeById(elementId);
      if (node) {
        const formKey = node.def.formKey;
        let schema: string = node.def.formKey;

        if (schema) {
          return {
            processName: processName,
            elementId: elementId,
            schema: schema,
          };
        } else {
          throw new InternalServerErrorException(
            `Cannot resolve schema from invalid formKey '${schema}', example: 'jsonschema://SimpleApproveReject' or 'custom://your-custom-key'`,
          );
        }
      } else {
        throw new NotFoundException(
          `Element ${elementId} not found in process ${processName}`,
        );
      }
    } catch (e) {
      throw new InternalServerErrorException(e);
    }
  }

  // helper
  workflowProcessArrayMaping(alldata: any[]) {
    const result = alldata.map((data) => this.workflowProcessMaping(data));
    return result;
  }
  workflowProcessMaping(data: any) {
    const result: WorkflowProcess = {
      id: data.id,
      name: data.name,
      status: data.status,
      startedAt: data.startedAt,
      items: data.items
        .filter((item) => item.type != 'bpmn:SequenceFlow')
        .map((item) => ({
          taskId: item.id,
          name: item.name,
          status: item.status,
          assignee: item.assignee,
          candidateUsers: item.candidateUsers,
          candidateGroups: item.candidateGroups,
          type: item.type,
        })),
    };
    return result;
  }

  /**
   * listen all events in bpmn-server engine, inclduded task and execution instances
   * @param event bpmn-server event instance
   */
  async applyEventListener(event) {
    if (event.context.item) {
      const workflowName: string = event.context.instance.name;
      const eventType: EXECUTION_EVENT = event.event;
      const elementType: BPMN_TYPE = event.context.item.element.type;
      const elementId: string = event.context.item.element.id;
      const elementName: string = event.context.item.element.name;
      const elementProps = event.context.item.element.def;
      // const usertaskinput = event.context.item.input ?? null;
      const data = event.context.instance.data;
      const vars = {};
      Object.assign(vars, event.context.item.input);
      event.context.item.input = {};
      const options = event.context.options;
      switch (elementType) {
        case BPMN_TYPE.UserTask:
          let usertaskeventtype: UserTaskEventType;
          if (eventType == EXECUTION_EVENT.node_start)
            usertaskeventtype = 'start';
          else if (eventType == EXECUTION_EVENT.node_assign)
            usertaskeventtype = 'assign';
          else if (eventType == EXECUTION_EVENT.node_end)
            usertaskeventtype = 'end';
          else if (eventType == EXECUTION_EVENT.transform_input)
            usertaskeventtype = 'invoked';
          else if (eventType == EXECUTION_EVENT.node_wait)
            usertaskeventtype = 'wait';
          // console.log("event-------=================================",eventType,usertaskeventtype)
          if (usertaskeventtype) {
            const props: UserTaskData = {
              workflowName: workflowName,
              eventType: usertaskeventtype,
              elementType: 'bpmn:UserTask',
              elementId: elementId,
              elementName: elementName,
              elementProps: {
                startedAt: event.context.item.startedAt,
                assignee: event.context.item.assignee,
                candidateGroups: event.context.item.candidateGroups,
                candidateUsers: event.context.item.candidateUsers,
                dueDate: event.context.item.dueDate,
                followUpDate: event.context.item.followUpDate,
                priority: event.context.item.priority,
                formKey: elementProps.formKey,
              },
              data: data,
              vars: vars,
              options: options,
            };
            this.eventEmitter.emit(
              `${workflowName}.${elementId}.${usertaskeventtype}`,
              props,
            );
          }

          break;
        case BPMN_TYPE.ServiceTask:
          let servicetaskeventtype: ServiceTaskEventType;
          if (eventType == EXECUTION_EVENT.node_start)
            servicetaskeventtype = 'start';
          else if (eventType == EXECUTION_EVENT.node_end)
            servicetaskeventtype = 'end';
          if (servicetaskeventtype) {
            const props: ServiceTaskData = {
              workflowName: workflowName,
              eventType: servicetaskeventtype,
              elementType: 'bpmn:ServiceTask',
              elementId: elementId,
              elementName: elementName,
              data: data,
              vars: vars,
              options: options,
            };
            this.eventEmitter.emit(
              `${workflowName}.${elementId}.${servicetaskeventtype}`,
              props,
            );
          }
          break;
        default: //do nothing at the moment
          break;
      }
    }
  }
}

```

**Reply by ralphhanna** [reply]
I have created new Repo, https://github.com/bpmnServer/bpmn-nest can you please add the code there 

The API is now stable, how can we change to match the API

**Comment by kstan79** [comment]
Inside is empty, i'm don't know what way you wish to do:
1. create empty nestjs project + integration work + some example?
`not recommend cause nestjs and plugins keep changing day by day, it obsolete immediately after few days of publish`
3. nestjs plugin which is work with current bpmn version?
`it is more recommended way but i still busy with my existing project, not yet able to fully test/implement bpmn-server in nest. Once core framework done, and few bpmn use case stable, I can assist to create the plugin.`

the purpose of previous message is to share with you implement `bpmn-server` in backend, we can use `event` as alternative solution when `import` is not suitable.
